File: app/src/main/java/com/sc/lesa/mediashar/jlib/P/VoicePlayThread.java
package com.sc.lesa.mediashar.jlib.P;

import android.media.AudioFormat;
import android.media.AudioManager;
import android.util.Log;

import com.sc.lesa.mediashar.jlib.io.VoicePack;
import com.sc.lesa.mediashar.jlib.media.AACDecoder;
import com.sc.lesa.mediashar.jlib.media.MyAudioTrack;
import com.sc.lesa.mediashar.jlib.server.Classifier;

public class VoicePlayThread extends Thread implements AACDecoder.OnDecodeDone {
    public final static String TAG = VoicePlayThread.class.getName();

    boolean exit;
    Classifier inputdata ;
    AACDecoder aacDecoder;
    MyAudioTrack myAudioTrack;
    boolean hasInitVoice;

    public VoicePlayThread(Classifier classifier){
        this.inputdata=classifier;
    }

    private void initVoiceDecoder(int ChannelMode,int EncodeFormat,int ChannelCount,
                                  int ByteRate,int SampleRate){
        int mChannelMode = ChannelMode==AudioFormat.CHANNEL_IN_MONO?
                AudioFormat.CHANNEL_OUT_MONO:AudioFormat.CHANNEL_OUT_STEREO;

        myAudioTrack = new MyAudioTrack(SampleRate,mChannelMode,
                EncodeFormat,AudioManager.STREAM_MUSIC);
        myAudioTrack.init();

        aacDecoder=new AACDecoder(ChannelCount,ByteRate,SampleRate);
        aacDecoder.setOnDecodeDone(this);
        aacDecoder.start();
    }

    @Override
    public void run() {
        while (!exit){

            VoicePack voicePack = (VoicePack) inputdata.getVoicePack();
            if (voicePack!=null){
                if (hasInitVoice==false){
                    initVoiceDecoder(voicePack.ChannelMode,voicePack.EncodeFormat,
                            voicePack.ChannelCount,voicePack.ByteRate,voicePack.SampleRate);
                    hasInitVoice=true;
                    try {
                        sleep(1000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    voicePack.presentationTimeUs=0;
                }
                try {
                    this.aacDecoder.decode(voicePack.datas, 0, voicePack.datas.length, voicePack.presentationTimeUs);
                }catch (Throwable e){
                    e.printStackTrace();
                }
                //LogUtil.appendToFile("/sdcard/tsstaac.aac",voicePack.datas);

            }
        }
        dirtory();
    }

    private void dirtory(){
        aacDecoder.stop();
        myAudioTrack.release();

        inputdata=null;
        aacDecoder=null;
        myAudioTrack=null;

        Log.i(TAG,"退出成功");
    }

    public void exit(){
        Log.i(TAG,"开始退出");
        this.exit=true;
    }

    @Override
    public void onDecodeData(byte[] bytes, int offset, int len) {
        this.myAudioTrack.playAudioTrack(bytes,offset,len);
    }

    @Override
    public void close() {

    }

}

File: app/src/main/java/com/sc/lesa/mediashar/jlib/threads/VoicePlayThread.kt
package com.sc.lesa.mediashar.jlib.threads

import android.media.AudioFormat
import android.media.AudioManager
import android.util.Log
import com.sc.lesa.mediashar.jlib.io.VoicePack
import com.sc.lesa.mediashar.jlib.media.AACDecoder
import com.sc.lesa.mediashar.jlib.media.AACDecoder.OnDecodeDone
import com.sc.lesa.mediashar.jlib.media.MyAudioTrack
import com.sc.lesa.mediashar.jlib.server.DataPackList

class VoicePlayThread(var inputdata: DataPackList) : Thread(TAG), OnDecodeDone {

    var exit = false
    lateinit var aacDecoder: AACDecoder
    lateinit var myAudioTrack: MyAudioTrack
    var hasInitVoice = false

    companion object {
        val TAG = VoicePlayThread::class.java.name
    }

    private fun initVoiceDecoder(ChannelMode: Int, EncodeFormat: Int, ChannelCount: Int,
                                 ByteRate: Int, SampleRate: Int) {
        val mChannelMode = if (ChannelMode == AudioFormat.CHANNEL_IN_MONO) AudioFormat.CHANNEL_OUT_MONO else AudioFormat.CHANNEL_OUT_STEREO
        myAudioTrack = MyAudioTrack(SampleRate, mChannelMode,
                EncodeFormat, AudioManager.STREAM_MUSIC)
        myAudioTrack.init()
        aacDecoder = AACDecoder(ChannelCount, ByteRate, SampleRate)
        aacDecoder.setOnDecodeDone(this)
        aacDecoder.init()
    }

    override fun run() {
        while (!exit) {
            val voicePack = inputdata.getVoicePack() as VoicePack?
            if (voicePack != null) {
                if (!hasInitVoice) {
                    initVoiceDecoder(voicePack.ChannelMode, voicePack.EncodeFormat,
                            voicePack.ChannelCount, voicePack.ByteRate, voicePack.SampleRate)
                    hasInitVoice = true

                    sleep(1000)
                    voicePack.presentationTimeUs = 0
                }
                try {
                    aacDecoder.decode(voicePack.datas, 0, voicePack.datas.size, voicePack.presentationTimeUs)
                } catch (e: Throwable) {
                    e.printStackTrace()
                }
                //LogUtil.appendToFile("/sdcard/tsstaac.aac",voicePack.datas);
            }
        }
        dirtory()
    }

    private fun dirtory() {
        aacDecoder.stop()
        myAudioTrack.release()
        Log.i(TAG, "退出成功")
    }

    fun exit() {
        Log.i(TAG, "开始退出")
        exit = true
    }

    override fun onDecodeData(bytes: ByteArray?, offset: Int, len: Int) {
        myAudioTrack.playAudioTrack(bytes, offset, len)
    }

    override fun onClose() {}

}